sol-anchor/idl.go

```diff
@@ -0,0 +1,612 @@
+	return t.Simple != ""
+}
+
+func (t *FieldType) IsSimplePubKey() bool {
+	return t.Simple == "publicKey"
+}
+
+func (t *FieldType) IsDefined() bool {
+	return t.Defined != ""
+}
+
+func (t *FieldType) IsArray() bool {
+	return t.Array != nil
+}
+
+func (t *FieldType) IsVec() bool {
+	return t.Vec != nil
+}
+
+func (t *FieldType) IsOption() bool {
+	return t.Option != nil
+}
+
+func (t *FieldType) Resolve() string {
+	if t.IsSimplePubKey() {
+		return "PubKey"
+	}
+
+	if t.IsSimple() {
+		return t.Simple
+	}
+
+	if t.IsDefined() {
+		return t.Defined
+	}
+
+	if t.IsArray() {
+		return t.Array.Type
+	}
+
+	if t.IsVec() {
+		return t.Vec.Type
+	}
+
+	if t.IsOption() {
+		if t.Option.IsSimple() {
+			return *t.Option.Simple
+		}
+
+		if t.Option.IsDefined() {
+			return *t.Option.Defined
+		}
+
+		if t.Option.IsVec() {
+			return t.Option.Vec.Type
+		}
+
+		if t.Option.IsArray() {
+			return t.Array.Type
+		}
+	}
+
+	return ""
+}
+
+func (t *FieldType) ResolveProtobufType() string {
+	return ToProtobufType(t.Resolve())
+}
+
+func PrintSimple(simpleType string, fieldName string, variableName string, isOptional bool) string {
+	fieldNameSnakeCase := toSnakeCase(fieldName, true)
+	fieldNameSnakeCaseWithoutInitialUnderscore := toSnakeCase(fieldName, false)
+
+	// If cast is needed
+	cast := CastInRustIfNeeded(simpleType)
+	if cast != "" {
+		if isOptional {
+			return fmt.Sprintf("%s: Some(%s.%s as %s),", fieldNameSnakeCaseWithoutInitialUnderscore, variableName, fieldNameSnakeCase, cast)
+		}
+		return fmt.Sprintf("%s: %s.%s as %s,", fieldNameSnakeCaseWithoutInitialUnderscore, variableName, fieldNameSnakeCase, cast)
+	}
+
+	// If cast is NOT needed
+	if isOptional {
+		return fmt.Sprintf("%s: Some(%s.%s),", fieldNameSnakeCaseWithoutInitialUnderscore, variableName, fieldNameSnakeCase)
+	}
+	return fmt.Sprintf("%s: %s.%s,", fieldNameSnakeCaseWithoutInitialUnderscore, variableName, fieldNameSnakeCase)
+}
+
+func PrintDefined(typeName string, fieldName string, variableName string, types []Type, isVec bool, isOption bool) string {
+	for _, t := range types {
+		if t.Name == typeName {
+			if t.Type.IsEnum() {
+				return fmt.Sprintf("%s: map_enum_%s(%s.%s),", toSnakeCase(fieldName, false), t.SnakeCaseName(), toSnakeCase(variableName, true), toSnakeCase(fieldName, true))
+			} else {
+				var fieldsInString strings.Builder
+				for _, structField := range t.Type.Struct.Fields {
+					fieldString := fmt.Sprintf("%s.%s", toSnakeCase(variableName, true), toSnakeCase(fieldName, true))
+					if variableName == "" {
+						fieldString = toSnakeCase(fieldName, true)
+					}
+
+					if isOption {
+						fieldString = fmt.Sprintf("%s: map_option_%s(%s.%s),", toSnakeCase(fieldName, true), toSnakeCase(fieldName, true), toSnakeCase(variableName, true), toSnakeCase(fieldName, true))
+						fieldsInString.WriteString(fieldString)
+						continue
+					}
+
+					fieldsInString.WriteString(structField.Type.Print(structField.Name, fieldString, types))
```

Comment:
make the 'print' method on 'structField' so you don't have to call structField.Name
It feels more natural to do `structField.Print(...)` than to print the Field from its type...
I feel that the abstractions are not all on the right objects and that they could be simplified, leading to cleaner code.
Also, methods like `PrintDefined` shouldn't be capitalized if they aren't used outside the package.